File: src/prediction/src/state_models.cpp
#include "state_models.hpp"
#include <iostream>
#include <stdexcept>

// CA Class Implementation
CA::CA(double dt) : dt(dt) {}

double CA::get_dt() const {
    return this->dt;
}

Eigen::RowVectorXd CA::step(const Eigen::RowVectorXd& x) const {
    double dt = this->dt;

    double px = x(0);
    double py = x(1);
    double v = x(2);
    double a = x(3);
    double theta = x(4);

    double cos_theta = std::cos(theta);
    double sin_theta = std::sin(theta);

    Eigen::RowVectorXd x_new(5);
    x_new << px + (v + 0.5 * a * dt) * cos_theta * dt,
             py + (v + 0.5 * a * dt) * sin_theta * dt,
             v + a * dt,
             a,
             theta;

    return x_new;
}

Eigen::RowVectorXd CA::H(const Eigen::RowVectorXd& x) const {
    // x = [x, y, v, a, theta]
    Eigen::RowVectorXd z(4);
    z << x(0), x(1), x(2), x(4);
    return z;
}

Eigen::MatrixXd CA::JA(const Eigen::RowVectorXd& x) const {
    double dt = this->dt;

    double px = x(0);
    double py = x(1);
    double v = x(2);
    double a = x(3);
    double theta = x(4);

    double cos_theta = std::cos(theta);
    double sin_theta = std::sin(theta);

    Eigen::MatrixXd JA_(5, 5);
    JA_ << 1, 0, cos_theta * dt, 0.5 * cos_theta * dt * dt, -(v + 0.5 * a * dt) * sin_theta * dt,
          0, 1, sin_theta * dt, 0.5 * sin_theta * dt * dt, (v + 0.5 * a * dt) * cos_theta * dt,
          0, 0, 1, dt, 0,
          0, 0, 0, 1, 0,
          0, 0, 0, 0, 1;

    return JA_;
}


Eigen::MatrixXd CA::JH(const Eigen::RowVectorXd& x) const {
    double dt = this->dt;

    Eigen::MatrixXd JH_(4, 5);
    JH_ << 1, 0, 0, 0, 0,
          0, 1, 0, 0, 0,
          0, 0, 1, 0, 0,
          0, 0, 0, 0, 1;

    return JH_;
}

// CTRA Class Implementation
CTRA::CTRA(double dt) : dt(dt) {}

double CTRA::get_dt() const {
    return this->dt;
}

Eigen::RowVectorXd CTRA::step(const Eigen::RowVectorXd& x) const {
    double dt = this->dt;

    double px = x(0);
    double py = x(1);
    double v = x(2);
    double a = x(3);
    double theta = x(4);
    double r = x(5);

    double cos_theta = std::cos(theta);
    double sin_theta = std::sin(theta);

    Eigen::RowVectorXd x_new(6);

    if (std::abs(r) > 0.1) {
        double theta_dt = theta + r * dt;
        double cos_theta_dt = std::cos(theta_dt);
        double sin_theta_dt = std::sin(theta_dt);

        x_new << px + v / r * (sin_theta_dt - sin_theta) + v / (r * r) * (cos_theta_dt + dt * r * sin_theta_dt - cos_theta),
                 py + v / r * (-cos_theta_dt + cos_theta) + v / (r * r) * (sin_theta_dt - dt * r * cos_theta_dt - sin_theta),
                 v + a * dt,
                 a,
                 theta + r * dt,
                 r;
    } else {
        x_new << px + v * cos_theta * dt,
                 py + v * sin_theta * dt,
                 v + a * dt,
                 a,
                 theta,
                 r;
    }

    return x_new;
}

Eigen::RowVectorXd CTRA::H(const Eigen::RowVectorXd& x) const {
    // x = [x, y, v, a, theta, theta_rate]
    Eigen::RowVectorXd z(4);
    z << x(0), x(1), x(2), x(4);
    return z;
}

Eigen::MatrixXd CTRA::JA(const Eigen::RowVectorXd& x) const {
    double dt = this->dt;

    double px = x(0);
    double py = x(1);
    double v = x(2);
    double a = x(3);
    double theta = x(4);
    double r = x(5);

    double cos_theta = std::cos(theta);
    double sin_theta = std::sin(theta);

    Eigen::MatrixXd JA_(6, 6);

    if (std::abs(r) > 0.1) {
        double theta_dt = theta + r * dt;
        double cos_theta_dt = std::cos(theta_dt);
        double sin_theta_dt = std::sin(theta_dt);

        JA_ << 1, 0, (sin_theta_dt - sin_theta) / r, (-cos_theta + cos_theta_dt + r * dt * sin_theta_dt) / (r * r), ((r * v + a * r * dt) * cos_theta_dt - a * sin_theta_dt - v * r * cos_theta + a * sin_theta) / (r * r), -2 / (r * r * r) * ((r * v + a * r * dt) * sin_theta_dt + a * cos_theta_dt - v * r * sin_theta - a * cos_theta) + ((v + a * dt) * sin_theta_dt + dt * (r * v + a * r * dt) * cos_theta_dt - dt * a * sin_theta_dt - v * sin_theta) / (r * r),
              0, 1, (-cos_theta_dt + cos_theta) / r, (-sin_theta + sin_theta_dt - r * dt * cos_theta_dt) / (r * r), ((r * v + a * r * dt) * sin_theta_dt + a * cos_theta_dt - v * r * sin_theta - a * cos_theta) / (r * r), -2 / (r * r * r) * ((-r * v - a * r * dt) * cos_theta_dt + a * sin_theta_dt + v * r * cos_theta - a * sin_theta) + ((-v - a * dt) * cos_theta_dt + dt * (r * v + a * r * dt) * sin_theta_dt + a * dt * cos_theta_dt + v * cos_theta) / (r * r),
              0, 0, 1, dt, 0, 0,
              0, 0, 0, 1, 0, 0,
              0, 0, 0, 0, 1, dt,
              0, 0, 0, 0, 0, 1;
    } else {
        JA_ << 1, 0, cos_theta * dt, 0.5 * cos_theta * dt * dt, -(v + 0.5 * a * dt) * sin_theta * dt, 0,
              0, 1, sin_theta * dt, 0.5 * sin_theta * dt * dt, (v + 0.5 * a * dt) * cos_theta * dt, 0,
              0, 0, 1, dt, 0, 0,
              0, 0, 0, 1, 0, 0,
              0, 0, 0, 0, 1, dt,
              0, 0, 0, 0, 0, 1;
    }

    return JA_;
}

Eigen::MatrixXd CTRA::JH(const Eigen::RowVectorXd& x) const {
    double dt = this->dt;

    double px = x(0);
    double py = x(1);
    double v = x(2);
    double a = x(3);
    double theta = x(4);
    double r = x(5);

    double cos_theta = std::cos(theta);
    double sin_theta = std::sin(theta);

    Eigen::MatrixXd JH_(4, 6);

    if (std::abs(r) > 0.1) {
        double theta_dt = theta + r * dt;
        double cos_theta_dt = std::cos(theta_dt);
        double sin_theta_dt = std::sin(theta_dt);

        JH_ << 1, 0, (sin_theta_dt - sin_theta) / r, (-cos_theta + cos_theta_dt + r * dt * sin_theta_dt) / (r * r), ((r * v + a * r * dt) * cos_theta_dt - a * sin_theta_dt - v * r * cos_theta + a * sin_theta) / (r * r), -2 / (r * r * r) * ((r * v + a * r * dt) * sin_theta_dt + a * cos_theta_dt - v * r * sin_theta - a * cos_theta) + ((v + a * dt) * sin_theta_dt + dt * (r * v + a * r * dt) * cos_theta_dt - dt * a * sin_theta_dt - v * sin_theta) / (r * r),
              0, 1, (-cos_theta_dt + cos_theta) / r, (-sin_theta + sin_theta_dt - r * dt * cos_theta_dt) / (r * r), ((r * v + a * r * dt) * sin_theta_dt + a * cos_theta_dt - v * r * sin_theta - a * cos_theta) / (r * r), -2 / (r * r * r) * ((-r * v - a * r * dt) * cos_theta_dt + a * sin_theta_dt + v * r * cos_theta - a * sin_theta) + ((-v - a * dt) * cos_theta_dt + dt * (r * v + a * r * dt) * sin_theta_dt + a * dt * cos_theta_dt + v * cos_theta) / (r * r),
              0, 0, 1, dt, 0, 0,
              0, 0, 0, 0, 1, dt;
    } else {
        JH_ << 1, 0, cos_theta * dt, 0.5 * cos_theta * dt * dt, -(v + 0.5 * a * dt) * sin_theta * dt, 0,
              0, 1, sin_theta * dt, 0.5 * sin_theta * dt * dt, (v + 0.5 * a * dt) * cos_theta * dt, 0,
              0, 0, 1, dt, 0, 0,
              0, 0, 0, 0, 1, dt;
    }

    return JH_;
}

File: src/prediction/src/state_models.py
#!/usr/bin/env python3
# coding: utf-8
import numpy as np

class CA():
    def __init__(self, dt=0.1):
        """
        CA 모델 클래스의 초기화 메서드

        Args:
            dt (float): 시간 간격 (기본값: 0.1)
        """
        self.dt = dt

    def step(self, x):
        """
        CA 모델의 한 단계 예측 수행

        Args:
            x (np.ndarray): 현재 상태 벡터 [x, y, v, a, theta]

        Returns:
            np.ndarray: 다음 시간 단계의 상태 벡터
        """
        dt = self.dt
        x_new = [x[0]+(x[2]+1/2*x[3]*dt)*np.cos(x[4])*dt,
                 x[1]+(x[2]+1/2*x[3]*dt)*np.sin(x[4])*dt,
                 x[2]+x[3]*dt,
                 x[3],
                 x[4]]

        return np.array(x_new)

    def H(self, x):
        """
        관측 모델

        Args:
            x (np.ndarray): 상태 벡터 [x, y, v, a, theta]

        Returns:
            np.ndarray: 관측 벡터
        """
        return np.array([x[0], x[1], x[2], x[4]])

    def JA(self, x, dt=0.1):
        """
        예측 모델의 야코비안

        Args:
            x (np.ndarray): 상태 벡터 [x, y, v, a, theta]
            dt (float): 시간 간격

        Returns:
            np.ndarray: 야코비안 행렬
        """
        px = x[0]
        py = x[1]
        v = x[2]
        a = x[3]
        yaw = x[4]

        JA_ = [[1, 0, np.cos(yaw)*dt, 1/2*np.cos(yaw)*dt**2, -(v+1/2*a*dt)*np.sin(yaw)*dt],
               [0, 1, np.sin(yaw)*dt, 1/2*np.sin(yaw)*dt**2,
                (v+1/2*a*dt)*np.cos(yaw)*dt],
               [0, 0, 1, dt, 0],
               [0, 0, 0, 1, 0],
               [0, 0, 0, 0, 1]]

        return np.array(JA_)

    def JH(self, x, dt=0.1):
        """
        관측 모델의 야코비안

        Args:
            x (np.ndarray): 상태 벡터 [x, y, v, a, theta]
            dt (float): 시간 간격

        Returns:
            np.ndarray: 야코비안 행렬
        """
        px = x[0]
        py = x[1]
        v = x[2]
        a = x[3]
        yaw = x[4]

        JH_ = np.array([[1, 0, 0, 0, 0],
                        [0, 1, 0, 0, 0],
                        [0, 0, 1, 0, 0],
                        [0, 0, 0, 0, 1]])

        return JH_


class CTRA():
    def __init__(self, dt=0.1):
        """
        CTRA 모델 클래스의 초기화 메서드

        Args:
            dt (float): 시간 간격 (기본값: 0.1)
        """
        self.dt = dt

    def step(self, x):
        """
        CTRA 모델의 한 단계 예측 수행

        Args:
            x (np.ndarray): 현재 상태 벡터 [x, y, v, a, theta, theta_rate]

        Returns:
            np.ndarray: 다음 시간 단계의 상태 벡터
        """
        if np.abs(x[5]) > 0.1:
            x_new = [x[0]+x[2]/x[5]*(np.sin(x[4]+x[5]*self.dt) -
                                     np.sin(x[4])) +
                     x[2]/(x[5]**2)*(np.cos(x[4]+x[5]*self.dt) +
                                     self.dt*x[5]*np.sin(x[4]+x[5]*self.dt) -
                                     np.cos(x[4])),
                     x[1]+x[2]/x[5]*(-np.cos(x[4]+x[5]*self.dt) +
                                     np.cos(x[4])) +
                     x[2]/(x[5]**2)*(np.sin(x[4]+x[5]*self.dt) -
                                     self.dt*x[5]*np.cos(x[4]+x[5]*self.dt) -
                                     np.sin(x[4])),
                     x[2]+x[3]*self.dt,
                     x[3],
                     x[4]+x[5]*self.dt,
                     x[5]]

        else:
            x_new = [x[0]+x[2]*np.cos(x[4])*self.dt,
                     x[1]+x[2]*np.sin(x[4])*self.dt,
                     x[2]+x[3]*self.dt,
                     x[3],
                     x[4],
                     x[5]]

        return np.array(x_new)

    def H(self, x):
        """
        관측 모델

        Args:
            x (np.ndarray): 상태 벡터 [x, y, v, a, theta, theta_rate]

        Returns:
            np.ndarray: 관측 벡터
        """
        return np.array([x[0], x[1], x[2], x[4]])

    def JA(self, x, dt=0.1):
        """
        예측 모델의 야코비안

        Args:
            x (np.ndarray): 상태 벡터 [x, y, v, a, theta, theta_rate]
            dt (float): 시간 간격

        Returns:
            np.ndarray: 야코비안 행렬
        """
        px = x[0]
        py = x[1]
        v = x[2]
        a = x[3]
        yaw = x[4]
        r = x[5]

        # upper
        if np.abs(r) > 0.1:
            JA_ = [[1, 0, (np.sin(yaw+r*dt)-np.sin(yaw))/r, (-np.cos(yaw)+np.cos(yaw+r*dt)+r*dt*np.sin(yaw+r*dt))/r**2,
                    ((r*v+a*r*dt)*np.cos(yaw+r*dt)-a*np.sin(yaw+r*dt) -
                     v*r*np.cos(yaw)+a*np.sin(yaw))/r**2,
                    -2/r**3*((r*v+a*r*dt)*np.sin(yaw+r*dt)+a*np.cos(yaw+r*dt)-v*r*np.sin(yaw)-a*np.cos(yaw)) +
                    ((v+a*dt)*np.sin(yaw+r*dt)+dt*(r*v+a*r*dt)*np.cos(yaw+r*dt)-dt*a*np.sin(yaw+r*dt)-v*np.sin(yaw))/r**2],
                   [0, 1, (-np.cos(yaw+r*dt)+np.cos(yaw))/r, (-np.sin(yaw)+np.sin(yaw+r*dt)-r*dt*np.cos(yaw+r*dt))/r**2,
                    ((r*v+a*r*dt)*np.sin(yaw+r*dt)+a*np.cos(yaw+r*dt) -
                     v*r*np.sin(yaw)-a*np.cos(yaw))/r**2,
                    -2/r**3*((-r*v-a*r*dt)*np.cos(yaw+r*dt)+a*np.sin(yaw+r*dt)+v*r*np.cos(yaw)-a*np.sin(yaw)) +
                    ((-v-a*dt)*np.cos(yaw+r*dt)+dt*(r*v+a*r*dt)*np.sin(yaw+r*dt)+a*dt*np.cos(yaw+r*dt)+v*np.cos(yaw))/r**2],
                   [0, 0, 1, dt, 0, 0],
                   [0, 0, 0, 1, 0, 0],
                   [0, 0, 0, 0, 1, dt],
                   [0, 0, 0, 0, 0, 1]]
        else:
            JA_ = [[1, 0, np.cos(yaw)*dt, 1/2*np.cos(yaw)*dt**2, -(v+1/2*a*dt)*np.sin(yaw)*dt, 0],
                   [0, 1, np.sin(yaw)*dt, 1/2*np.sin(yaw)*dt**2,
                    (v+1/2*a*dt)*np.cos(yaw)*dt, 0],
                   [0, 0, 1, dt, 0, 0],
                   [0, 0, 0, 1, 0, 0],
                   [0, 0, 0, 0, 1, dt],
                   [0, 0, 0, 0, 0, 1]]

        return np.array(JA_)

    def JH(self, x, dt=0.1):
        """
        관측 모델의 야코비안

        Args:
            x (np.ndarray): 상태 벡터 [x, y, v, a, theta, theta_rate]
            dt (float): 시간 간격

        Returns:
            np.ndarray: 야코비안 행렬
        """
        px = x[0]
        py = x[1]
        v = x[2]
        a = x[3]
        yaw = x[4]
        r = x[5]

        # upper
        if np.abs(r) > 0.1:

            JH_ = [[1, 0, (np.sin(yaw+r*dt)-np.sin(yaw))/r, (-np.cos(yaw)+np.cos(yaw+r*dt)+r*dt*np.sin(yaw+r*dt))/r**2,
                    ((r*v+a*r*dt)*np.cos(yaw+r*dt)-a*np.sin(yaw+r*dt) -
                     v*r*np.cos(yaw)+a*np.sin(yaw))/r**2,
                    -2/r**3*((r*v+a*r*dt)*np.sin(yaw+r*dt)+a*np.cos(yaw+r*dt)-v*r*np.sin(yaw)-a*np.cos(yaw)) +
                    ((v+a*dt)*np.sin(yaw+r*dt)+dt*(r*v+a*r*dt)*np.cos(yaw+r*dt)-dt*a*np.sin(yaw+r*dt)-v*np.sin(yaw))/r**2],
                   [0, 1, (-np.cos(yaw+r*dt)+np.cos(yaw))/r, (-np.sin(yaw)+np.sin(yaw+r*dt)-r*dt*np.cos(yaw+r*dt))/r**2,
                    ((r*v+a*r*dt)*np.sin(yaw+r*dt)+a*np.cos(yaw+r*dt) -
                     v*r*np.sin(yaw)-a*np.cos(yaw))/r**2,
                    -2/r**3*((-r*v-a*r*dt)*np.cos(yaw+r*dt)+a*np.sin(yaw+r*dt)+v*r*np.cos(yaw)-a*np.sin(yaw)) +
                    ((-v-a*dt)*np.cos(yaw+r*dt)+dt*(r*v+a*r*dt)*np.sin(yaw+r*dt)+a*dt*np.cos(yaw+r*dt)+v*np.cos(yaw))/r**2],
                   [0, 0, 1, dt, 0, 0],
                   [0, 0, 0, 0, 1, dt]]

        else:
            JH_ = [[1, 0, np.cos(yaw)*dt, 1/2*np.cos(yaw)*dt**2, -(v+1/2*a*dt)*np.sin(yaw)*dt, 0],
                   [0, 1, np.sin(yaw)*dt, 1/2*np.sin(yaw)*dt**2,
                    (v+1/2*a*dt)*np.cos(yaw)*dt, 0],
                   [0, 0, 1, dt, 0, 0],
                   [0, 0, 0, 0, 1, dt]]

        return np.array(JH_)

    def pred(self, x, t_pred):
        """
        주어진 시간만큼 예측하여 상태 벡터 목록 반환

        Args:
            x (np.ndarray): 초기 상태 벡터 [x, y, v, a, theta, theta_rate]
            t_pred (float): 예측할 시간

        Returns:
            np.ndarray: 예측된 상태 벡터들로 이루어진 목록
        """
        self.x = x

        x_list = [self.x]
        for t in range(int(t_pred/self.dt)):
            x_list.append(self.step(self.x))

        return np.array(x_list)
